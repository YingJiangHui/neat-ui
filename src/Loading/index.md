---
nav:
  title: Components
  path: /components
---

## Loading

### 基础使用

```tsx
import React from 'react';
import { Loading } from 'neat-ui-react';

export default () => <Loading />;
```

### 大小

```tsx
import React from 'react';
import { Loading } from 'neat-ui-react';

export default () => (
  <>
    <Loading size="small" />
    <br />
    <br />
    <Loading size="middle" />
    <br />
    <br />
    <Loading size="large" />
  </>
);
```

### 颜色

```tsx
import React from 'react';
import { Loading } from 'neat-ui-react';

export default () => (
  <>
    <Loading color="#0070f3" />
    <br />
    <br />
    <Loading color="#fc011a" />
    <br />
    <br />
    <Loading color="#ffce44" />
  </>
);
```

<API src="loading.tsx"></API>

### Loading Container

对于 Container 中的内容加载时

```tsx
import React, { useState } from 'react';
import { Loading } from 'neat-ui-react';

export default () => {
  const [loading, setLoading] = useState(false);
  return (
    <>
      <Loading.Container loading={loading}>
        <p>这是一段内容</p>
      </Loading.Container>
      <button onClick={() => setLoading((loading) => !loading)}>off</button>
    </>
  );
};
```

<API src="loading-container.tsx"></API>
