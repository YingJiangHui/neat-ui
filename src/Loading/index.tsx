import React, { FC, HTMLAttributes, PropsWithChildren } from 'react';

const defaultProps = {};

type LoadingProps = {};
type Props = typeof defaultProps & LoadingProps & HTMLAttributes<any>;
const Loading: FC<PropsWithChildren<Props>> = ({ ...rest }) => {
  return (
    <div className="loading-container">
      <div className="loading">
        <i />
        <i />
        <i />
      </div>
      <style jsx>{`
        .loading {
          > i {
            margin-right: 10px;
            display: inline-block;
            width: 0.8em;
            height: 0.8em;
            background: #dadada;
            border-radius: 50%;
            animation: loading-zoom 1.4s infinite both;
          }

          > i:nth-child(2) {
            animation-delay: 0.2s;
          }

          > i:nth-child(3) {
            animation-delay: 0.4s;
          }
        }

        @keyframes loading-zoom {
          0% {
            opacity: 0.2;
          }
          20% {
            opacity: 1;
          }
          100% {
            opacity: 0.2;
          }
        }
      `}</style>
    </div>
  );
};

export default Loading;
