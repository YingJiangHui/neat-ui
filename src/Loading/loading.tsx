import React, { FC, HTMLAttributes, PropsWithChildren, useMemo } from 'react';
import withDefaults from '@/utils/with-defaults';
import LoadingContainer from '@/Loading/loading-container';
import { useTheme } from '@/hooks/use-theme';

interface LoadingProps {
  size?: 'small' | 'middle' | 'large';
  color?: string;
}

const defaultProps = {
  size: 'middle',
};
type Props = PropsWithChildren<
  typeof defaultProps & LoadingProps & HTMLAttributes<any>
>;
const sizeMap = {
  small: 4,
  middle: 6,
  large: 8,
};
const Loading: FC<Props> = ({ size, color, children, ...rest }) => {
  const width = useMemo(() => sizeMap[size], [size]);
  const theme = useTheme();
  return (
    <>
      <div className="loading-container" {...rest}>
        <div className="loading">
          <i />
          <i />
          <i />
        </div>
      </div>
      <style jsx>{`
        .loading > i {
          width: ${width}px;
          height: ${width}px;
          background: ${color || theme.palette.grayscale_8};
        }
      `}</style>
      <style jsx>{`
        .loading-container {
          display: inline-block;
        }

        .loading {
          display: flex;
          align-items: center;
        }

        .loading > i {
          margin-right: 2px;
          display: inline-block;
          border-radius: 50%;
          animation: loading-zoom 1.4s infinite both;
        }

        .loading > i:nth-child(2) {
          animation-delay: 0.2s;
        }

        .loading > i:nth-child(3) {
          animation-delay: 0.4s;
          margin-right: 0;
        }

        @keyframes loading-zoom {
          0% {
            opacity: 0.2;
          }
          20% {
            opacity: 1;
          }
          100% {
            opacity: 0.2;
          }
        }
      `}</style>
    </>
  );
};

type LoadingComponent<P = {}> = React.FC<P> & {
  Container: typeof LoadingContainer;
};
export default withDefaults(
  Loading,
  defaultProps,
) as LoadingComponent<LoadingProps>;
