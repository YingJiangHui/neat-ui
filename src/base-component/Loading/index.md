---
nav:
  title: Components
  path: /components
---

## Loading

Demo:

```tsx
import React from 'react';
import { Loading } from 'neat-ui-react';

export default () => <Loading />;
```
