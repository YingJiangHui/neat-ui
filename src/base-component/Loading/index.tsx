import React, { FC, HTMLAttributes, PropsWithChildren } from 'react';
import './index.less';
import '@/vars.less';
import styled from '@emotion/styled';
const defaultProps = {};

type LoadingProps = {};

const LoadingStyle = styled.ul`
  list-style: none;
  > li {
    margin-right: 10px;
    display: inline-block;
    width: 1em;
    height: 1em;
    background: #dadada;
    border-radius: 50%;
  }
`;

type Props = typeof defaultProps & LoadingProps & HTMLAttributes<any>;
const Loading: FC<PropsWithChildren<Props>> = ({ ...rest }) => {
  return (
    <LoadingStyle {...rest}>
      {[0.3, 0.6, 0.9].map((timeout) => (
        <li style={{ animation: `zoom infinite .8s linear ${timeout}s` }} />
      ))}
    </LoadingStyle>
  );
};

export default Loading;
