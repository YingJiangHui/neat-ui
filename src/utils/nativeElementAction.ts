import { unmountComponentAtNode } from 'react-dom';

export const unmountElement = (container: HTMLElement | null) => {
  // 退出时进行清理
  unmountComponentAtNode(container as HTMLDivElement);
  container!.remove();
  container = null;
};

export const mountElement = () => {
  const div = document.createElement('div');
  document.body.appendChild(div);
  return div;
};
