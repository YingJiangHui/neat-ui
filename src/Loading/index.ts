import LoadingContainer from './loading-container';
import Loading from './loading';
Loading.Container = LoadingContainer;
export { Loading };
