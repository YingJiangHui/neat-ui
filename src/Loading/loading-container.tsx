import React, { FC, HTMLAttributes, PropsWithChildren } from 'react';
import withDefaults from '@/utils/with-defaults';
import Loading from './loading';
import classnames from '@/shared/classnames';

type LoadingContainerProps = {
  indicator?: React.ReactNode;
  loading?: boolean;
  opacity?: number;
  maskColor: string;
};
const defaultProps: LoadingContainerProps = {
  indicator: <Loading />,
  loading: false,
  opacity: 0.5,
  maskColor: '#fff',
};
type Props = PropsWithChildren<
  typeof defaultProps & LoadingContainerProps & HTMLAttributes<any>
>;
const LoadingContainer: FC<PropsWithChildren<Props>> = ({
  loading,
  indicator,
  children,
  opacity,
  maskColor,
  ...rest
}) => {
  return (
    <div className="loading-container-wrapper">
      {loading && <div className="loading-mask">{indicator}</div>}
      <div
        className={classnames('loading-container', loading && 'loading-blur')}
        {...rest}
      >
        {children}
      </div>
      <style jsx>{`
        .loading-container-wrapper {
          position: relative;
        }

        .loading-mask {
          overflow: hidden;
          position: absolute;
          left: 0;
          top: 0;
          width: 100%;
          height: 100%;
          display: flex;
          align-items: center;
          justify-content: center;
        }

        .loading-container {
          transition: opacity 0.25s;
        }

        .loading-blur {
          opacity: 0.5;
        }

        .loading-blur::after {
          content: '';
          position: absolute;
          top: 0;
          right: 0;
          bottom: 0;
          left: 0;
          width: 100%;
          height: 100%;
          background: ${maskColor};
          transition: all 0.25s;
          z-index: 10;
          opacity: ${opacity};
        }
      `}</style>
    </div>
  );
};
export default withDefaults(LoadingContainer, defaultProps);
