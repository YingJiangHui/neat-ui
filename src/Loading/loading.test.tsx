import renderer from 'react-test-renderer';
import { Loading } from '@/Loading/index';
import { act } from 'react-dom/test-utils';
import React from 'react';
import { render } from 'react-dom';

import { mountElement, unmountElement } from '@/utils/nativeElementAction';
let container: null | HTMLDivElement = null;

describe('Loading.Container', () => {
  it('loading快照', () => {
    const component = renderer.create(<Loading />);
    expect(component.toJSON()).toMatchSnapshot();
  });
});

describe('Loading.Container', () => {
  beforeEach(() => {
    container = mountElement();
    console.log(container);
  });
  afterEach(unmountElement.bind(null, container));
  it('可以遮罩盖住内容', () => {
    act(() => {
      render(<Loading.Container loading={true} />, container as HTMLElement);
    });
    expect(container?.querySelector('.loading-blur')).toBeTruthy();
  });
});
