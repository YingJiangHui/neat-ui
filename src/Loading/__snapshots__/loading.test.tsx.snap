// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Loading.Container loading快照 1`] = `
<div
  className="loading-container"
>
  <div
    className="loading"
  >
    <i />
    <i />
    <i />
  </div>
  <style>
    
          .loading &gt; i {
            width: 14px;
            height: 14px;
            margin-right: 9.333333333333334px;
          }
        
  </style>
  <style>
    
          .loading &gt; i {
            display: inline-block;
            background: #dadada;
            border-radius: 50%;
            animation: loading-zoom 1.4s infinite both;
          }

          .loading &gt; i:nth-child(2) {
            animation-delay: 0.2s;
          }

          .loading &gt; i:nth-child(3) {
            animation-delay: 0.4s;
            margin-right: 0;
          }

          @keyframes loading-zoom {
            0% {
              opacity: 0.2;
            }
            20% {
              opacity: 1;
            }
            100% {
              opacity: 0.2;
            }
          }
        
  </style>
</div>
`;
